src/utils/allocators.zig

```diff
@@ -177,121 +209,191 @@ pub const DiskMemoryAllocator = struct {
             file.handle,
             0,
         ) catch |err| {
-            std.debug.print("Disk Memory Allocator error: {}\n", .{err});
+            self.logFailure(err, file_name);
             return null;
         };
 
-        return memory.ptr;
-    }
-
-    /// unmaps the memory (file still exists and is removed on deinit())
-    pub fn free(_: *anyopaque, buf: []u8, log2_align: u8, return_address: usize) void {
-        _ = log2_align;
-        _ = return_address;
-        // TODO: build a mapping from ptr to file so we can delete the corresponding file on free
-        const buf_aligned_len = std.mem.alignForward(usize, buf.len, std.mem.page_size);
-        std.posix.munmap(@alignCast(buf.ptr[0..buf_aligned_len]));
+        std.debug.assert(size <= file_aligned_size - @sizeOf(Metadata)); // sanity check
+        std.mem.bytesAsValue(Metadata, full_alloc[size..][0..@sizeOf(Metadata)]).* = .{
+            .file_index = file_index,
+        };
+        return full_alloc.ptr;
     }
 
-    /// not supported rn
     fn resize(
-        _: *anyopaque,
-        buf_unaligned: []u8,
-        log2_buf_align: u8,
+        ctx: *anyopaque,
+        buf: []u8,
+        log2_align: u8,
         new_size: usize,
         return_address: usize,
     ) bool {
-        // not supported
-        _ = buf_unaligned;
-        _ = log2_buf_align;
-        _ = new_size;
         _ = return_address;
-        return false;
+        const self: *Self = @ptrCast(@alignCast(ctx));
+
+        const alignment = @as(usize, 1) << @intCast(log2_align);
+        std.debug.assert(alignment <= std.mem.page_size); // the allocator interface shouldn't allow this (aside from the *Raw methods).
+
+        const old_file_aligned_size = alignedFileSize(buf.len);
+        const old_mmap_aligned_size = alignedMmapSize(old_file_aligned_size, self.mmap_ratio);
+
+        const new_file_aligned_size = alignedFileSize(new_size);
+        const new_mmap_aligned_size = alignedMmapSize(new_file_aligned_size, self.mmap_ratio);
+
+        if (new_mmap_aligned_size > old_mmap_aligned_size) {
+            return false;
+        }
+
+        if (new_file_aligned_size == old_file_aligned_size) {
+            return true;
+        }
+
+        const buf_ptr: [*]align(std.mem.page_size) u8 = @alignCast(buf.ptr);
+        const metadata: Metadata = @bitCast(buf_ptr[old_file_aligned_size - @sizeOf(Metadata) ..][0..@sizeOf(Metadata)].*);
+
+        const file_name_bounded = fileNameBounded(metadata.file_index);
+        const file_name = file_name_bounded.constSlice();
+
+        const file = self.dir.openFile(file_name, .{ .mode = .read_write }) catch |err| {
+            self.logFailure(err, file_name);
+            return false;
+        };
+        defer file.close();
+
+        file.setEndPos(new_file_aligned_size) catch return false;
+        if (new_mmap_aligned_size < old_mmap_aligned_size) {
+            // we'll never be able to recover this anyway, because the `alignedFileSize` will
```

Comment:
The way I refactored it after our conversation was to mmap more memory based on the mmap_ratio in the `alloc` call, which then grants more available but uncommitted memory.

If we allow shrinking the allocation such that the mmap size we derive from it becomes smaller, the rest of the address space becomes unobtainable, since any resizes which would require growing the mmap size is rejected. So keeping the memory around with this design would simply be a potential waste, and possibly even hide out of bounds accesses on many item ptrs by causing the uncommitted memory to be committed vs causing a segmentation fault.

I'm open to adding the real size as metadata to avoid this path if you want, that just comes with the tradeoff of committing more pages earlier the bigger an allocation becomes (`@sizeOf(Metadata)` becomes bigger) - not by a huge amount obviously, but still a tradeoff.

Reply:
1308c9ef293afb5b3428b1fa8f3887d1ac827318 (see other comment)